src/actix/auth.rs

```diff
@@ -126,33 +126,42 @@ where
             return Box::pin(self.service.call(req));
         }
 
-        // Grab API key from request
-        let key =
-            // Request header
-            req.headers().get("api-key").and_then(|key| key.to_str().ok()).map(|key| key.to_string())
-                // Fall back to authentication header with bearer token
-                .or_else(|| {
-                    Authorization::<Bearer>::parse(&req).ok().map(|auth| auth.as_ref().token().into())
-                });
-
-        if let Some(key) = key {
-            let is_allowed = if let Some(ref auth_keys) = self.auth_keys {
-                auth_keys.can_write(&key) || (is_read_only(&req) && auth_keys.can_read(&key))
-            } else {
-                // This code path should not be reached
-                log::warn!("Auth for REST API is set up incorrectly. Denying access by default.");
-                false
-            };
-            if is_allowed {
+        let invalid_api_key = |req: ServiceRequest| -> Self::Future {
+            Box::pin(async {
+                Ok(req
+                    .into_response(HttpResponse::Forbidden().body("Invalid api-key"))
+                    .map_into_right_body())
+            })
+        };
+
+        let auth_keys = &self.auth_keys;
+
+        let api_key = req
+            .headers()
+            .get("api-key")
+            .and_then(|key| key.to_str().ok());
+
+        let bearer_key = Authorization::<Bearer>::parse(&req);
+        let bearer_key = bearer_key.as_ref().ok().map(|auth| auth.as_ref().token());
+
+        if let Some(key) = api_key.or(bearer_key) {
```

Comment:
Can we lazily parse the bearer key? such that we don't parse it if we have a valid api key from the `"api-key"` header.

Reply:
I think the flow should be like this: 
If the api-key is incorrect, we should return an error right away. Otherwise, check the bearer token, first as an api-key, and if it fails, then as a JWT (or the other way around, open for discussion).

Reply:
I don't see a clean way to do it without duplicating code or using clones. The original version were using `to_string()` and `into()` to clone strings. Mine operates on `&str`.

```diff
-        // Grab API key from request
-        let key =
-            // Request header
-            req.headers().get("api-key").and_then(|key| key.to_str().ok()).map(|key| key.to_string())
-                // Fall back to authentication header with bearer token
-                .or_else(|| {
-                    Authorization::<Bearer>::parse(&req).ok().map(|auth| auth.as_ref().token().into())
-                });
```